File: app/src/main/java/com/VegaSolutions/lpptransit/lppapi/responseobjects/GzipInterceptor.java
package com.VegaSolutions.lpptransit.lppapi.responseobjects;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;

import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okhttp3.internal.http.RealResponseBody;
import okio.BufferedSink;
import okio.GzipSink;
import okio.GzipSource;
import okio.Okio;

public class GzipInterceptor implements Interceptor {
    @NotNull
    @Override
    public Response intercept(@NotNull Chain chain) throws IOException {

        Request request = chain.request();
        Response response = chain.proceed(request);

        if (response.body() != null && isGzipped(response)) {
            long contentLength = response.body().contentLength();
            GzipSource responseBody = new GzipSource(response.body().source());
            Headers strippedHeaders = response.headers().newBuilder().build();
            return response.newBuilder().headers(strippedHeaders)
                    .body(new RealResponseBody(response.body().contentType().toString(), contentLength, Okio.buffer(responseBody)))
                    .build();
        }

        return response;

    }

    private boolean isGzipped(Response response) {
        String encoding = response.header("Content-Encoding");
        return encoding != null && encoding.equals("gzip");
    }
}

File: app/src/main/java/com/VegaSolutions/lpptransit/data/lppapi/GzipInterceptor.kt
package com.VegaSolutions.lpptransit.data.lppapi

import okhttp3.Interceptor
import okhttp3.Response
import okhttp3.internal.http.RealResponseBody
import okio.GzipSource
import okio.buffer
import java.io.IOException

class GzipInterceptor : Interceptor {
    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        val response = chain.proceed(request)

        val body = response.body ?: return response
        if (response.header("Content-Encoding") != "gzip") return response

        val contentLength = body.contentLength()
        val responseBody = GzipSource(body.source())
        val strippedHeaders = response.headers.newBuilder().build()
        return response.newBuilder().headers(strippedHeaders)
            .body(
                RealResponseBody(
                    body.contentType().toString(),
                    contentLength,
                    responseBody.buffer()
                )
            )
            .build()
    }

}
